### PR TITLE
[FLINK-25129][docs]project configuation changes in docs

### PR DESCRIPTION
**Pull request by infoverload**
## What is the purpose of the change

A unified configuration section for both DataStream users and Table API users.  

## Brief change log

- configuration pages split up into multiple pages and put into its own section under Application Development

## Verifying this change

Please make sure both new and modified tests in this PR follows the conventions defined in our code quality guide: https://flink.apache.org/contributing/code-style-and-quality-common.html#testing

This change is already covered by existing tests, such as *(please describe tests)*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)

### CODE REVIEW

File: docs/content/docs/connectors/table/kafka.md
Comment (slinkydeveloper) [inline docs/content/docs/connectors/table/kafka.md:41]: This sentence should be in the other sql connectors as well, except filesystem, which is included by default

File: docs/content/docs/dev/configuration/overview.md
Comment (slinkydeveloper) [inline docs/content/docs/dev/configuration/overview.md:88]: Perhaps this goes before the explanation on provided? Between the maven snippet and the "important" text block

File: docs/content/docs/dev/configuration/overview.md
Comment (slinkydeveloper) [inline docs/content/docs/dev/configuration/overview.md:86]: Actually, this sounds like already the next step and might be build system dependant (not sure how it works with gradle honestly)

File: docs/content/docs/dev/configuration/overview.md
Comment (slinkydeveloper) [inline docs/content/docs/dev/configuration/overview.md:97]: Perhaps here add the links to maven/gradle/sbt pages?

**Review by slinkydeveloper — COMMENTED** [review]
Did a quick scan now, looks good. Will add later some new phrasing and explanations.

I wonder, why a separate section with several pages more than a single page?

File: docs/content/docs/dev/configuration/connector.md
Comment (slinkydeveloper) [inline docs/content/docs/dev/configuration/connector.md:65]: I start suggesting some wording:

> When loading uber jars directly in the distribution, you lock on a specific version of the connector, simplifying the management of a shared multi-job flink cluster, but as a downside each job developer doesn't have control over the connector version. While when shading the connector jar, either the thin or the fat one, in your job JAR, guarantees each job developer more control over the connector version. In case of shading the thin JAR, the job developer has even control over the transitive dependencies, as it can bump them without bumping the connector (binary compatibility permitting).

@fapaul @twalthr wanna add some details here on why one should use thin vs uber jars?

**Review by slinkydeveloper — CHANGES_REQUESTED** [review]

File: docs/content/docs/dev/configuration/advanced.md
Comment (MartijnVisser) [inline docs/content/docs/dev/configuration/advanced.md:47]: This probably needs to be both DataStream and Table API

File: docs/content/docs/dev/configuration/advanced.md
Comment (MartijnVisser) [inline docs/content/docs/dev/configuration/advanced.md:68]: Flink doesn't support Scala 2.11 anymore since the next release

File: docs/content/docs/dev/configuration/advanced.md
Comment (MartijnVisser) [inline docs/content/docs/dev/configuration/advanced.md:72]: I would say only use Flink's Java APIs and Flink's Scala APIs

File: docs/content/docs/dev/configuration/advanced.md
Comment (MartijnVisser) [inline docs/content/docs/dev/configuration/advanced.md:100]: Flink Scala API

File: docs/content/docs/dev/configuration/advanced.md
Comment (MartijnVisser) [inline docs/content/docs/dev/configuration/advanced.md:141]: interactions

**Review by MartijnVisser — COMMENTED** [review]
Had some first comments but it looks really good already. I'll try to finish this evening or else on Monday

File: docs/content/docs/connectors/datastream/cassandra.md
Comment (MartijnVisser) [inline docs/content/docs/connectors/datastream/cassandra.md:40]: NB: All these links currently send me to http://localhost:1313/flink/flink-docs-master/docs/dev/configuration/. That page is blank, it should be http://localhost:1313/flink/flink-docs-master/docs/dev/configuration/overview/ I think. So:

```suggestion
Note that the streaming connectors are currently __NOT__ part of the binary distribution. See how to link with them for cluster execution [here]({{< ref "docs/dev/configuration/overview" >}}).
```

File: docs/content/docs/dev/configuration/maven.md
Comment (MartijnVisser) [inline docs/content/docs/dev/configuration/maven.md:1]: I'm leaning towards moving the content of Maven, Gradle and SBT all to sections under the Overview. Reasons for that is that if something changes on the Project Configuration, you can immediately change it in one go and keep things in sync. Probably with the tabs option to see different setups (a tab for a Maven example, tab for Gradle etc). Something like that?

What made you think about splitting them to separate pages?

**Review by MartijnVisser — CHANGES_REQUESTED** [review]
Reviewed the last part, found one broken link and curious to your thoughts on my other comment

File: docs/content/docs/dev/configuration/maven.md
Comment (infoverload) [reply docs/content/docs/dev/configuration/maven.md:1]: In the first iteration, I thought that it made the page too long and that too much scrolling makes for bad UX, but I can try the tabs idea.

Comment (MartijnVisser) [reply docs/content/docs/dev/configuration/maven.md:1]: That's a fair point but I'm also not sure if we provide perfect support for both Gradle and SBT. I would lean towards integrating those in one page.

Comment (slinkydeveloper) [reply docs/content/docs/dev/configuration/maven.md:1]: @MartijnVisser you mean something like this? https://github.com/apache/flink/pull/18353#discussion_r789423489

File: docs/content/docs/dev/configuration/advanced.md
Comment (slinkydeveloper) [inline docs/content/docs/dev/configuration/advanced.md:109]: This is very table related. Perhaps name this paragraph "Anatomy of table dependencies"? and then you make the next paragraph _Table Planner and Table Planner Loader_ a subparagraph of this one?

File: docs/content/docs/dev/configuration/advanced.md
Comment (slinkydeveloper) [inline docs/content/docs/dev/configuration/advanced.md:101]: Replace _This has now_ with _Since 1.15_

File: docs/content/docs/dev/configuration/advanced.md
Comment (slinkydeveloper) [inline docs/content/docs/dev/configuration/advanced.md:107]: This sentence is incorrect, I think i badly worded it initially. What I was trying to explain is that while for java apis, everything is already in the distribution so the dependency can be "provided" (as explained in the overview and maven page), for scala apis you don't have the jar built-in in the distribution, so you need to package the jar somehoe, effectively like a connector dependency.

Perhaps reword this to something like:

> While Table Java APIs artifacts are built-in in the distribution, Table Scala API artifacts are not included by default. You need to either include them in the distribution `/lib` folder manually (recommended), or package them as dependencies in your job uber jar.

File: docs/content/docs/connectors/table/upsert-kafka.md
Comment (slinkydeveloper) [inline docs/content/docs/connectors/table/upsert-kafka.md:52]: remove _currently_, as it implies that at some point in the future it will be part of the distribution. Same for other changes to connectors pages

File: docs/content/docs/dev/table/overview.md
Comment (slinkydeveloper) [inline docs/content/docs/dev/table/overview.md:44]: I would rather go and remove these tabs all together. Perhaps just link to the project conf page for Java and Scala and the python overview page for Python? @MartijnVisser any opinions here?

**Review by slinkydeveloper — COMMENTED** [review]
I think it looks good, I have mostly minor comments and I also asked the review from other people regarding the various build scripts and new wording we are introducing with this PR.

File: docs/content/docs/dev/table/overview.md
Comment (MartijnVisser) [reply docs/content/docs/dev/table/overview.md:44]: I would +1 removal of these tabs here. The overview page for Table should be comparable to the current DataStream Overview, so an overview of the Table APIs capabilities etc. We've moved the setup to the project configuration page, which fits better. Python is kind of a weird one with its own requirements, so for now better to link to the Python. 

So instead of having a section Table Program Dependencies, perhaps just a link to the project configuration page with a text like "Check out the project configuration for setting up your environment so you can start building a Table API program"? (But then in nicer phrasing 😅 )

**Review by MartijnVisser — APPROVED** [review]
I think content-wise this is ready to merge. @slinkydeveloper what do you think?

I do think you need to squash commits on your end. But not 100% sure on the actual requirements on this.

File: docs/content/docs/dev/configuration/gradle.md
Comment (matriv) [reply docs/content/docs/dev/configuration/gradle.md:25]: I've opened 2 PRs to ugprade to 7.3.3: 
https://github.com/apache/flink-web/pull/504
https://github.com/apache/flink/pull/18609

**Comment by slinkydeveloper** [comment]
There are still some unresolved comments, once they're solved we're ready to merge

File: docs/content/docs/dev/configuration/gradle.md
Comment (matriv) [inline docs/content/docs/dev/configuration/gradle.md:104]: @infoverload Please state here the `installDist`

File: docs/content/docs/dev/configuration/gradle.md
Comment (matriv) [inline docs/content/docs/dev/configuration/gradle.md:106]: And here the installShadowDist

File: docs/content/docs/dev/configuration/overview.md
Comment (matriv) [inline docs/content/docs/dev/configuration/overview.md:138]: Here and next line it should be `implementation` instead of `compile`.

File: docs/content/docs/dev/configuration/overview.md
Comment (matriv) [inline docs/content/docs/dev/configuration/overview.md:145]: Here and the next 3 lines it should be `runtimeOnly` instead of `compile`.

**Review by matriv — APPROVED** [review]
Maybe we could still provide a link to https://nightlies.apache.org/flink/flink-docs-release-1.14/docs/dev/datastream/project-configuration/#gradle, where there is also the script to create a new project with the samples, but this is for a streaming job and the `build.gradle` file is a bit different, so I'll leave it up to you.

LGTM from my side.

File: docs/content/docs/dev/configuration/gradle.md
Comment (zentol) [inline docs/content/docs/dev/configuration/gradle.md:39]: Which project folder and files? Which project should be imported?
Are we missing a reference to the gradle quickstart?

File: docs/content/docs/dev/configuration/maven.md
Comment (zentol) [inline docs/content/docs/dev/configuration/maven.md:40]: same issue as with the gradle guide.

File: docs/content/docs/dev/configuration/overview.md
Comment (zentol) [inline docs/content/docs/dev/configuration/overview.md:118]: We shouldn't point users to snapshot dependencies.

File: docs/content/docs/dev/configuration/overview.md
Comment (zentol) [inline docs/content/docs/dev/configuration/overview.md:139]: @matriv This is different than what you propose in https://github.com/apache/flink-web/pull/504/files.

File: docs/content/docs/dev/configuration/overview.md
Comment (zentol) [inline docs/content/docs/dev/configuration/overview.md:164]: Is this actually necessary?

File: docs/content/docs/dev/configuration/overview.md
Comment (zentol) [inline docs/content/docs/dev/configuration/overview.md:44]: This is oddly specific, technically incorrect (you need way more than flink-runtime), it shouldn't contain a hard-coded version, and the link is quite weird (why not maven central?)

File: docs/content/docs/dev/configuration/maven.md
Comment (zentol) [inline docs/content/docs/dev/configuration/maven.md:145]: This should be unnecessary because all Flink dependencies have a provided dependency on slf4j and log4j.
Also, the log4j groupId is outdated (`org.apache.logging.log4j` would be the correct choice).

File: docs/content/docs/dev/configuration/maven.md
Comment (zentol) [inline docs/content/docs/dev/configuration/maven.md:133]: this is different than the version in our quickstarts.

File: docs/content/docs/dev/configuration/maven.md
Comment (zentol) [inline docs/content/docs/dev/configuration/maven.md:127]: Overall I find it weird that we don't just link to the quickstarts. They are ultimately the source of truth of how a recommended project looks like, and are also tested.

File: docs/content/docs/dev/configuration/maven.md
Comment (zentol) [inline docs/content/docs/dev/configuration/maven.md:161]: This is missing the `ServicesResourceTransformer`.

File: docs/content/docs/dev/configuration/overview.md
Comment (zentol) [inline docs/content/docs/dev/configuration/overview.md:104]: As is this property is unused, and none of the instructions on this page reference it.

File: docs/content/docs/dev/configuration/overview.md
Comment (zentol) [inline docs/content/docs/dev/configuration/overview.md:99]: `myorg.org` points to an actively used domain. I would suggest to just remove it.

File: docs/content/docs/dev/configuration/advanced.md
Comment (zentol) [inline docs/content/docs/dev/configuration/advanced.md:40]: rt.jar doesn't exist in JDK 9+. I would just remove the references to any jars.

**Review by zentol — CHANGES_REQUESTED** [review]

File: docs/content/docs/dev/configuration/overview.md
Comment (matriv) [reply docs/content/docs/dev/configuration/overview.md:118]: But if they want `1.15-SNAPSHOT` ?

File: docs/content/docs/dev/configuration/overview.md
Comment (matriv) [reply docs/content/docs/dev/configuration/overview.md:139]: It's a mistake I did there, I'll fix it. thx!

File: docs/content/docs/dev/configuration/overview.md
Comment (zentol) [reply docs/content/docs/dev/configuration/overview.md:118]: Then they need to figure out how to use it. Users are not the target audience for snapshot artifacts; in fact we are not allowed to advertise such artifacts.

Comment (matriv) [reply docs/content/docs/dev/configuration/overview.md:118]: Ok, because in the docs/examples we are using the <<flinkVersion>> which for master is `1.15-SNAPSHOT`

File: docs/content/docs/dev/configuration/maven.md
Comment (slinkydeveloper) [reply docs/content/docs/dev/configuration/maven.md:127]: I think this page is more for detailed/advanced topics for maven. This is a bit duplicated but I think it should be synced with what we have in the quickstarts. We can also improve it later after this PR.

Comment (infoverload) [reply docs/content/docs/dev/configuration/maven.md:127]: After some discussion, I think we can improve it after the PR.

File: docs/content/docs/dev/configuration/overview.md
Comment (zentol) [inline docs/content/docs/dev/configuration/overview.md]: these 2 properties were used though ;)

File: docs/content/docs/dev/configuration/overview.md
Comment (zentol) [inline docs/content/docs/dev/configuration/overview.md]: Why was this removed?

Comment (MartijnVisser) [reply docs/content/docs/dev/configuration/overview.md]: We talked to @tillrohrmann about it - The SBT example was still targeting 0.13.13 while SBT 1.6.1 is the latest version. We decided to first remove it and create a follow-up ticket for adding an SBT example later.

Comment (infoverload) [reply docs/content/docs/dev/configuration/overview.md]: I had already removed the standalone page on SBT and had forgotten to remove this section.

**Pull request merged**
